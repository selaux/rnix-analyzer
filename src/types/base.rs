use super::{NixType, TypeId};
use id_arena::Arena;

pub struct BaseTypes {
    pub unknown: TypeId,
    pub string: TypeId,
    pub path: TypeId,
    pub integer: TypeId,
    pub float: TypeId,
    pub boolean: TypeId,
    pub null: TypeId,
}

impl BaseTypes {
    pub fn register(type_arena: &mut Arena<NixType>) -> Self {
        let unknown = type_arena.alloc(NixType::Unknown);
        let string = type_arena.alloc(NixType::String);
        let path = type_arena.alloc(NixType::Path);
        let integer = type_arena.alloc(NixType::Integer);
        let float = type_arena.alloc(NixType::Float);
        let boolean = type_arena.alloc(NixType::Boolean);
        let null = type_arena.alloc(NixType::Null);
        BaseTypes {
            unknown,
            string,
            path,
            integer,
            float,
            boolean,
            null,
        }
    }
}
