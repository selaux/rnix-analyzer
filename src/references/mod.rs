use crate::{scope::defines_scope, DefinitionId, Scope, ScopeKind, Scopes};
use id_arena::{Arena, Id as ArenaId};
use rnix::types::{AttrSet, BinOp, Ident, Inherit, TokenWrapper, TypedNode};
use rnix::{SyntaxKind, SyntaxNode, TextRange, WalkEvent, AST};
use std::collections::{BTreeMap, VecDeque};

type IdentifierId = ArenaId<Identifier>;

#[derive(Debug, PartialEq, Eq, Clone, Hash)]
pub struct Identifier {
    pub id: IdentifierId,
    pub name: String,
    pub text_range: TextRange,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum ReferenceError {
    NotFoundInScope(String, TextRange),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Reference {
    from: IdentifierId,
    to: DefinitionId,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct References {
    identifier_arena: Arena<Identifier>,
    pub references: BTreeMap<IdentifierId, Reference>,
}

fn filter_identifier(
    parents: &VecDeque<SyntaxNode>,
    current_attrset_recursive: &VecDeque<bool>,
    node: &SyntaxNode,
) -> bool {
    let last_four_parents: Vec<_> = (1..4).map(|v| parents.get(v)).collect();
    let last_four_kinds: Vec<_> = last_four_parents
        .iter()
        .map(|n| n.map(|n| n.kind()))
        .collect();

    // Filter identifiers in select paths (except first one)
    if (last_four_kinds[0] == Some(SyntaxKind::NODE_SELECT)
        || last_four_kinds[0] == Some(SyntaxKind::NODE_KEY))
        && node.prev_sibling_or_token().map(|prev| prev.kind()) == Some(SyntaxKind::TOKEN_DOT)
    {
        return true;
    }
    // Filter identifiers in normal attrset paths
    if last_four_kinds[0] == Some(SyntaxKind::NODE_KEY)
        && last_four_kinds[1] == Some(SyntaxKind::NODE_KEY_VALUE)
        && last_four_kinds[2] == Some(SyntaxKind::NODE_ATTR_SET)
        && !current_attrset_recursive[0]
    {
        return true;
    }
    // Filter identifiers in from clause of inherit
    if last_four_kinds[0] == Some(SyntaxKind::NODE_INHERIT) {
        let parent = last_four_parents[0]
            .cloned()
            .and_then(Inherit::cast)
            .and_then(|inherit| inherit.from());
        if parent.is_some() {
            return true;
        }
    }
    // Filter identifiers in object contains binary operator
    if last_four_kinds[0] == Some(SyntaxKind::NODE_BIN_OP) {
        let parent = last_four_parents[0].cloned().and_then(BinOp::cast);
        if let Some(parent) = parent {
            let im_right = parent
                .rhs()
                .map(|n| node.text_range().is_subrange(&n.text_range()))
                .unwrap_or(false);
            if im_right {
                return true;
            }
        }
    }
    if last_four_kinds[0] == Some(SyntaxKind::NODE_SELECT)
        && last_four_kinds[1] == Some(SyntaxKind::NODE_BIN_OP)
    {
        let parent = last_four_parents[1].cloned().and_then(BinOp::cast);
        if let Some(parent) = parent {
            let im_right = parent
                .rhs()
                .map(|n| node.text_range().is_subrange(&n.text_range()))
                .unwrap_or(false);
            if im_right {
                return true;
            }
        }
    }
    false
}

impl References {
    pub fn from_ast_and_scope_tree(ast: &AST, scopes: &Scopes) -> (Self, Vec<ReferenceError>) {
        let mut errors = vec![];
        let mut references = BTreeMap::new();
        let mut identifier_arena = Arena::new();
        let mut current_attrset_recursive: VecDeque<bool> = VecDeque::new();
        let mut parents: VecDeque<SyntaxNode> = VecDeque::new();
        let mut in_scopes: VecDeque<&Scope> = VecDeque::new();

        in_scopes.push_front(scopes.root_scope());

        for walk_event in ast.node().preorder() {
            match walk_event {
                WalkEvent::Enter(node) => {
                    let kind = node.kind();
                    parents.push_front(node.clone());
                    if defines_scope(kind) {
                        let scope = scopes
                            .scope_arena
                            .iter()
                            .map(|v| v.1)
                            .find(|scope| {
                                scope.text_range == node.text_range()
                                    && scope.kind != ScopeKind::Root
                            })
                            .expect("scope should exist");
                        in_scopes.push_front(scope);
                    }

                    match kind {
                        SyntaxKind::NODE_ATTR_SET => {
                            let recursive = AttrSet::cast(node)
                                .map(|attrset| attrset.recursive())
                                .unwrap_or(false);
                            current_attrset_recursive.push_front(recursive);
                        }
                        SyntaxKind::NODE_IDENT => {
                            if !filter_identifier(&parents, &current_attrset_recursive, &node) {
                                if let Some(ident) = Ident::cast(node) {
                                    let name = ident.as_str().to_owned();
                                    let id = identifier_arena.alloc_with_id(|id| Identifier {
                                        id,
                                        name: name.clone(),
                                        text_range: ident.node().text_range(),
                                    });
                                    let definition = in_scopes
                                        .iter()
                                        .filter_map(|scope| scope.get_definition(&name))
                                        .next();
                                    if let Some(definition) = definition {
                                        references.insert(
                                            id,
                                            Reference {
                                                from: id,
                                                to: definition,
                                            },
                                        );
                                    } else {
                                        let non_exhaustive =
                                            in_scopes.iter().any(|scope| !scope.is_exhaustive());
                                        if !non_exhaustive {
                                            errors.push(ReferenceError::NotFoundInScope(
                                                name,
                                                ident.node().text_range(),
                                            ));
                                        }
                                    }
                                }
                            }
                        }
                        _ => {}
                    }
                }
                WalkEvent::Leave(node) => {
                    let kind = node.kind();
                    if kind == SyntaxKind::NODE_ATTR_SET {
                        current_attrset_recursive.pop_front();
                    }
                    if defines_scope(kind) {
                        in_scopes.pop_front();
                    }
                    parents.pop_front();
                }
            }
        }

        (
            References {
                identifier_arena,
                references,
            },
            errors,
        )
    }

    /// Returns the applicable scopes for a given text range
    pub fn get_identifiers(&self) -> Vec<&Identifier> {
        self.identifier_arena.iter().map(|(_id, val)| val).collect()
    }
}

#[cfg(test)]
mod tests {
    // Note this useful idiom: importing names from outer (for mod tests) scope.
    use super::*;
    use crate::{AnalysisError, AnalysisResult};
    use insta::{assert_debug_snapshot, assert_display_snapshot};
    use std::process::Command;
    use std::str;

    #[test]
    fn test_references_simple_function() {
        assert_refences_snapshot("a: a");
    }

    #[test]
    fn test_references_function_pattern() {
        assert_refences_snapshot("{a,...}: a");
    }

    #[test]
    fn test_references_let_in() {
        assert_refences_snapshot("let a = 1; in a");
    }

    #[test]
    fn test_references_let_in_object() {
        assert_refences_snapshot("let a.b = 1; in a.b");
    }

    #[test]
    fn test_references_object_contains() {
        assert_refences_snapshot("let a.b = 1; in a ? c");
    }

    #[test]
    fn test_references_object_contains_path() {
        assert_refences_snapshot("let a.b = 1; in a ? c.d");
    }

    #[test]
    fn test_references_inherit_from() {
        assert_refences_snapshot("let a.b = 1; in { inherit (a) b; }");
    }

    #[test]
    fn test_references_object() {
        assert_refences_snapshot(
            "b: {
            a = 1;
            \"${b}\" = 2;
            c.d = 3;
        }",
        );
    }

    #[test]
    fn test_references_rec_object() {
        assert_refences_snapshot(
            "b: rec {
            a = 1;
            \"${b}\" = 2;
            c.d = a;
        }",
        );
    }

    #[test]
    fn test_error_undefined_reference_simple_function() {
        assert_refences_error_snapshot("a: b");
    }

    #[test]
    fn test_undefined_reference_inside_with() {
        assert_refences_snapshot("a: with a; b");
    }

    fn assert_refences_snapshot(nix_code: &str) {
        let output = Command::new("nix")
            .args(&["eval", &format!("({})", nix_code)])
            .output()
            .expect("failed to run nix");
        println!("{}", str::from_utf8(&output.stdout).unwrap());
        eprintln!("{}", str::from_utf8(&output.stderr).unwrap());
        assert_eq!(output.status.code(), Some(0));
        let parse_result = rnix::parse(nix_code);
        assert_eq!(parse_result.errors(), vec![]);
        let result = AnalysisResult::from(&parse_result);
        let errors: Vec<AnalysisError> = result.errors().cloned().collect();
        assert_eq!(errors, vec![]);
        let references = References::from_ast_and_scope_tree(&parse_result, &result.scopes);
        assert_eq!(references.1, vec![]);
        assert_display_snapshot!(format!(
            "{}\n=========\n{:#?}\n=========\n{:#?}",
            nix_code,
            references.0.get_identifiers(),
            references.0.references
        ));
    }

    fn assert_refences_error_snapshot(nix_code: &str) {
        let output = Command::new("nix")
            .args(&["eval", &format!("({})", nix_code)])
            .output()
            .expect("failed to run nix");
        println!("{}", str::from_utf8(&output.stdout).unwrap());
        eprintln!("{}", str::from_utf8(&output.stderr).unwrap());
        assert_eq!(output.status.code(), Some(1));
        let parse_result = rnix::parse(nix_code);
        assert_eq!(parse_result.errors(), vec![]);
        let result = AnalysisResult::from(&parse_result);
        let references = References::from_ast_and_scope_tree(&parse_result, &result.scopes);
        assert_debug_snapshot!(references.1);
    }
}
