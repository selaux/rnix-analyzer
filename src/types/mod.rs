use crate::{utils::Stack, CollectFromTree, DefinitionId, Reference, Scope, VariableId};
use id_arena::{Arena, Id as ArenaId};
use rnix::{types::ParsedType, value::Value, SyntaxKind};
use std::collections::{BTreeMap, BTreeSet, HashMap};
use std::convert::TryFrom;
use std::hash::Hash;
use std::iter::FromIterator;

mod base;

use base::BaseTypes;

/// Unique Id of a type
pub type TypeId = ArenaId<NixType>;

pub type TypeError = ();

#[derive(Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
pub enum NixType {
    Unknown,
    String,
    Path,
    Integer,
    Float,
    Boolean,
    Null,
    List(TypeId),
    AttrSet {
        attrs: BTreeMap<String, TypeId>,
        exhaustive: bool,
    },
    Lambda {
        arguments: BTreeMap<String, TypeId>,
        at: TypeId,
        returns: TypeId,
    },
    Union(BTreeSet<TypeId>),
}

impl NixType {
    pub(crate) fn to_type_descriptor(
        &self,
        type_arena: &Arena<NixType>,
        current_indent: usize,
    ) -> String {
        match self {
            NixType::Unknown => "unknown".to_owned(),
            NixType::String => "string".to_owned(),
            NixType::Path => "path".to_owned(),
            NixType::Integer => "int".to_owned(),
            NixType::Float => "float".to_owned(),
            NixType::Boolean => "bool".to_owned(),
            NixType::Null => "null".to_owned(),
            NixType::List(type_id) => format!(
                "list({})",
                type_arena[*type_id].to_type_descriptor(type_arena, current_indent)
            ),
            NixType::Union(types) => {
                let mut v: Vec<String> = types
                    .iter()
                    .map(|ty| type_arena[*ty].to_type_descriptor(type_arena, current_indent))
                    .collect();
                v.sort();
                v.join(" | ")
            }
            // TODO
            NixType::AttrSet { .. } => "set()".to_owned(),
            NixType::Lambda { .. } => "lambda()".to_owned(),
        }
    }
}

impl Default for NixType {
    fn default() -> Self {
        NixType::Unknown
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct Types {
    type_arena: Arena<NixType>,
    type_for_definition: BTreeMap<DefinitionId, TypeId>,
    export_type: TypeId,
}

impl Types {
    pub fn nixtype_description(&self, type_id: &TypeId) -> Option<String> {
        self.nixtype(type_id)
            .map(|t| t.to_type_descriptor(&self.type_arena, 0))
    }

    /// Returns a scope by id
    pub fn nixtype(&self, id: &TypeId) -> Option<&NixType> {
        self.type_arena.get(*id)
    }

    /// Returns a scope by id
    pub fn export_nixtype(&self) -> &TypeId {
        &self.export_type
    }
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) struct ListContext {
    types: Vec<NixType>,
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) enum TrackTypesContext {
    Root,
    None,
    List(ListContext),
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) struct TrackTypesDependencies<'a, 'b> {
    references: &'a BTreeMap<VariableId, Reference>,
    current_scopes: &'b Stack<Scope>,
}

impl TrackTypesDependencies<'_, '_> {
    pub fn new<'a, 'b>(
        references: &'a BTreeMap<VariableId, Reference>,
        current_scopes: &'b Stack<Scope>,
    ) -> TrackTypesDependencies<'a, 'b> {
        TrackTypesDependencies {
            references,
            current_scopes,
        }
    }
}

pub(crate) struct TrackTypesState {
    type_arena: Arena<NixType>,
    type_hashes: HashMap<NixType, TypeId>,
    base_types: BaseTypes,
    type_for_definition: BTreeMap<DefinitionId, TypeId>,
    export_type: TypeId,
    contexts: Stack<TrackTypesContext>,
}

impl TrackTypesState {
    fn get_or_insert_type(&mut self, typ: NixType) -> TypeId {
        if let Some(typ) = self.type_hashes.get(&typ) {
            *typ
        } else {
            self.type_arena.alloc(typ)
        }
    }
}

impl Default for TrackTypesState {
    fn default() -> Self {
        let mut type_arena = Arena::new();
        let base_types = BaseTypes::register(&mut type_arena);
        let export_type = base_types.unknown;

        TrackTypesState {
            type_arena,
            type_hashes: HashMap::new(),
            base_types,
            type_for_definition: BTreeMap::new(),
            export_type,
            contexts: Stack::default(),
        }
    }
}

#[derive(Default)]
pub(crate) struct TrackTypes {
    state: TrackTypesState,
}

impl TrackTypes {
    pub fn new() -> Self {
        Default::default()
    }
}

impl<'a, 'b> CollectFromTree<TrackTypesDependencies<'a, 'b>> for TrackTypes {
    type State = TrackTypesState;
    type Result = Types;
    type Error = TypeError;

    fn enter_node(&mut self, _: TrackTypesDependencies<'a, 'b>, node: &rnix::SyntaxNode) {
        println!("Enter {:?}", node.kind());
        match node.kind() {
            SyntaxKind::NODE_ROOT => self.state.contexts.push_front(TrackTypesContext::Root),
            SyntaxKind::NODE_LIST => self
                .state
                .contexts
                .push_front(TrackTypesContext::List(ListContext { types: vec![] })),
            _ => self.state.contexts.push_front(TrackTypesContext::None),
        }
    }

    fn exit_node(&mut self, _: TrackTypesDependencies<'a, 'b>, node: &rnix::SyntaxNode) {
        let state = &mut self.state;
        let context = state.contexts.pop_front();
        println!("Exit {:?}", node.kind());
        let parsed_type = ParsedType::try_from(node.clone());
        let current_type_id = match parsed_type {
            Ok(ParsedType::Root(_)) => state.base_types.unknown,
            Ok(ParsedType::Value(v)) => match v.to_value() {
                Ok(Value::Float(_)) => state.base_types.float,
                Ok(Value::Integer(_)) => state.base_types.integer,
                Ok(Value::Path(_, _)) => state.base_types.path,
                Ok(Value::String(_)) => state.base_types.string,
                _ => unreachable!(),
            },
            Ok(ParsedType::Str(_)) => state.base_types.string,
            Ok(ParsedType::List(_)) => match context {
                Some(TrackTypesContext::List(context)) => {
                    let types_hashset = BTreeSet::from_iter(context.types.into_iter());
                    let ids: BTreeSet<_> = types_hashset
                        .into_iter()
                        .map(|typ| state.get_or_insert_type(typ))
                        .collect();
                    let type_id = match ids.len() {
                        0 => state.base_types.unknown,
                        1 => ids.into_iter().next().unwrap(),
                        _ => state.get_or_insert_type(NixType::Union(ids)),
                    };
                    state.get_or_insert_type(NixType::List(type_id))
                }
                c => {
                    log::warn!("Exited list node, but found non-list-context: {:?}, returning list(unknown) type", c);
                    state.get_or_insert_type(NixType::List(state.base_types.unknown))
                }
            },
            Ok(_) => {
                log::warn!(
                    "Found non-inferrable node {:?}, returning unknown type",
                    node
                );
                state.base_types.unknown
            }
            Err(_) => {
                log::warn!(
                    "Error parsing into ParsedType for node {:?}, returning unknown type",
                    node
                );
                state.base_types.unknown
            }
        };
        let current_type = &state.type_arena[current_type_id];
        let parent_context = state.contexts.get_mut(0);
        match parent_context {
            Some(TrackTypesContext::Root) => state.export_type = current_type_id,
            Some(TrackTypesContext::List(parent_context)) => {
                parent_context.types.push(current_type.clone());
            }
            _ => {}
        }
    }

    fn state(&self) -> &Self::State {
        &self.state
    }

    fn result(self) -> (Self::Result, Vec<Self::Error>) {
        (
            Types {
                type_arena: self.state.type_arena,
                type_for_definition: self.state.type_for_definition,
                export_type: self.state.export_type,
            },
            vec![],
        )
    }
}

#[cfg(test)]
mod tests {
    use crate::{AnalysisError, AnalysisResult};
    use insta::assert_display_snapshot;
    use std::process::Command;
    use std::str;

    #[test]
    fn test_integer_export_nixtype() {
        assert_export_nixtype_snapshot("1");
    }

    #[test]
    fn test_float_export_nixtype() {
        assert_export_nixtype_snapshot("1.0");
    }

    #[test]
    fn test_path_export_nixtype() {
        assert_export_nixtype_snapshot("./.gitignore");
    }

    #[test]
    fn test_string_export_nixtype() {
        assert_export_nixtype_snapshot("\"\"");
    }

    #[test]
    fn test_empty_array_export_nixtype() {
        assert_export_nixtype_snapshot("[]");
    }

    #[test]
    fn test_integer_array_export_nixtype() {
        assert_export_nixtype_snapshot("[ 1 2 3 ]");
    }

    #[test]
    fn test_mixed_array_export_nixtype() {
        assert_export_nixtype_snapshot("[ 1 2.0 \"3\" \"1\" 2 3.0]");
    }

    fn assert_export_nixtype_snapshot(nix_code: &str) {
        let output = Command::new("nix")
            .args(&["eval", &format!("({})", nix_code)])
            .output()
            .expect("failed to run nix");
        println!("{}", str::from_utf8(&output.stdout).unwrap());
        eprintln!("{}", str::from_utf8(&output.stderr).unwrap());
        assert_eq!(output.status.code(), Some(0));
        let parse_result = rnix::parse(nix_code);
        assert_eq!(parse_result.errors(), vec![]);
        let result = AnalysisResult::from(&parse_result);
        let errors: Vec<AnalysisError> = result.errors().cloned().collect();
        assert_eq!(errors, vec![]);
        let export_nixtype = result.export_nixtype();
        let description = result
            .nixtype_description(export_nixtype)
            .expect("should be there");
        assert_display_snapshot!(format!("{}\n=========\n{}", nix_code, description));
    }
}
