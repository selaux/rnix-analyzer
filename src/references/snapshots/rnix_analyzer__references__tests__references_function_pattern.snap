---
source: src/references/mod.rs
expression: references.0
---
References {
    identifiers: [
        Identifier {
            name: "a",
            text_range: [1; 2),
        },
        Identifier {
            name: "a",
            text_range: [9; 10),
        },
    ],
    references: {
        Identifier {
            name: "a",
            text_range: [1; 2),
        }: Reference {
            from: Identifier {
                name: "a",
                text_range: [1; 2),
            },
            to: Definition {
                id: 0,
                name: "a",
                text_range: [1; 2),
            },
        },
        Identifier {
            name: "a",
            text_range: [9; 10),
        }: Reference {
            from: Identifier {
                name: "a",
                text_range: [9; 10),
            },
            to: Definition {
                id: 0,
                name: "a",
                text_range: [1; 2),
            },
        },
    },
}
